api/lib/opentelemetry/trace/link.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Trace
+    # A link to a {Span}. Used (for example) in batching operations, where a
+    # single batch handler processes multiple requests from different traces.
+    # A Link can be also used to reference spans from the same trace. A Link
+    # and its attributes are immutable.
+    class Link
+      EMPTY_ATTRIBUTES = {}.freeze
+
+      # Returns the {SpanContext} for this link
+      #
+      # @return [SpanContext]
+      attr_reader :context
+
+      # Returns a new link.
+      #
+      # @param [SpanContext] span_context The context of the linked {Span}.
+      # @param [optional Hash<String, Object>] attributes A hash of attributes
+      #   for this link. Attributes will be frozen during Link initialization.
+      #   If both attributes and a link_formatter are provided, an ArgumentError
+      #   will be raised.
+      # @yieldreturn [optional Hash<String, Object>] attribute_formatter A
+      #   callable that returns a hash of attributes for this link. Will be
+      #   called lazily and its return value will be frozen when attributes are
+      #   first accessed. If both attributes and a link_formatter are provided,
+      #   an ArgumentError will be raised.
+      # @return [Link]
+      # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+      def initialize(span_context:, attributes: nil, &attribute_formatter)
+        raise ArgumentError unless span_context.instance_of?(SpanContext)
+        raise ArgumentError unless attributes.nil? || attributes.is_a?(Hash)
+        raise ArgumentError unless attributes.nil? || Internal.valid_attributes?(attributes)
```

Comment:
Maybe move the `attributes.is_a?(Hash)` into `valid_attributes` and then replace these two lines with:
```ruby
raise ArgumentError unless Internal.valid_attributes?(attributes)
```
(you already have the nil check there).